### PR TITLE
Remove all checking for the gametype in the game

### PR DESCRIPTION
**Pull request by heinrich5991**
OK, maybe not actually remove because it is kept for fallback when the
new method isn't available.

The whole gametype parsing business had the same downsides as user agent
parsing on the web, hence I removed it while keeping behavior the same.

This allows servers to explicitly opt in or out of certain bug
workarounds and other client behavior. This increases the complexity of
different configurations that are available in the client (which is a
bad thing).

### CODE REVIEW

File: datasrc/network.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 				  "NO_HAMMER_HIT", "NO_SHOTGUN_HIT", "NO_GRENADE_HIT", "NO_RIFLE_HIT", "NO_HOOK",
 				  "TELEGUN_GUN", "TELEGUN_GRENADE", "TELEGUN_LASER",
 				  "WEAPON_HAMMER", "WEAPON_GUN", "WEAPON_SHOTGUN", "WEAPON_GRENADE", "WEAPON_LASER", "WEAPON_NINJA"]
-GameInfoFlags = ["TIMESCORE"]
+GameInfoFlags = ["TIMESCORE", "GAMETYPE_RACE", "GAMETYPE_FASTCAP", "GAMETYPE_FNG", "GAMETYPE_DDRACE", "GAMETYPE_DDNET", "GAMETYPE_BLOCK_WORLDS", "GAMETYPE_VANILLA", "GAMETYPE_PLUS", "FLAG_STARTS_RACE", "RACE", "UNLIMITED_AMMO", "DDRACE_RECORD_MESSAGE", "RACE_RECORD_MESSAGE", "ALLOW_EYE_WHEEL", "ALLOW_HOOK_COLL", "ALLOW_ZOOM", "BUG_DDRACE_GHOST", "BUG_DDRACE_INPUT", "BUG_FNG_LASER_RANGE", "BUG_VANILLA_BOUNCE", "PREDICT_FNG", "PREDICT_DDRACE", "PREDICT_DDRACE_TILES", "PREDICT_VANILLA", "ENTITIES_DDNET", "ENTITIES_DDRACE", "ENTITIES_RACE", "ENTITIES_FNG", "ENTITIES_VANILLA"]
```
Comment (Learath2) [inline datasrc/network.py]: multilinepls

File: datasrc/network.py [hunk 2]
```diff
@@ -216,8 +221,9 @@
 		NetIntRange("m_AuthLevel", "AUTHED_NO", "AUTHED_ADMIN"),
 	]),
 
-	NetObjectEx("DDNetGameInfo", "gameinfo@netobj.ddnet.tw", [
+	NetObjectEx("GameInfoEx", "gameinfo@netobj.ddnet.tw", [
```
Comment (fokkonaut) [inline datasrc/network.py]: DDNetGameInfo was a good name, it matches the DDNetPlayer and DDNetCharacter.

File: src/game/server/gamecontroller.cpp [hunk 3]
```diff
@@ -535,11 +535,28 @@ void IGameController::Snap(int SnappingClient)
...
+		| GAMEINFOFLAG_ALLOW_EYE_WHEEL
+		| GAMEINFOFLAG_ALLOW_HOOK_COLL
+		| GAMEINFOFLAG_ALLOW_ZOOM
+		| GAMEINFOFLAG_BUG_DDRACE_GHOST
+		| GAMEINFOFLAG_BUG_DDRACE_INPUT
+		| GAMEINFOFLAG_PREDICT_DDRACE
+		| GAMEINFOFLAG_PREDICT_DDRACE_TILES
+		| GAMEINFOFLAG_ENTITIES_DDNET
+		| GAMEINFOFLAG_ENTITIES_DDRACE
+		| GAMEINFOFLAG_ENTITIES_RACE;
```
[the remark this answers is not in the document]
Comment (heinrich5991) [reply src/game/server/gamecontroller.cpp]: This was meant as "it's useful to display the entities as DDNet, DDRace, or Race, use the most sophisticated one you have".

File: datasrc/network.py (hunk 2 above)
Comment (heinrich5991) [reply datasrc/network.py]: I don't want this solely for DDNet though, but rather also for Race game types, etc.

**Comment by heinrich5991** [comment]
Still need to fix that the smaller gameinfo object is simply dropped by the network code.

**Comment by def-** [comment]
@heinrich5991 Ran into a conflict. Should I rebase and submit?

File: src/game/client/components/mapimages.h [hunk 4]
```diff
@@ -11,7 +11,7 @@ class CMapImages : public CComponent
 	int m_aTextures[64];
 	int m_Count;
 
-	char m_aEntitiesGameType[16];
+	const char *m_pEntitiesGameType;
```
Comment (Learath2) [inline src/game/client/components/mapimages.h]: How safe is this? You are assigning a pointer to a string literal to they have static storage duration like in C?

File: src/game/client/gameclient.cpp [hunk 5]
```diff
@@ -953,6 +950,98 @@ void CGameClient::ProcessEvents()
...
 }
 
+static CGameInfo GetGameInfo(const CNetObj_GameInfoEx *pInfoEx, int InfoExSize, CServerInfo *pFallbackServerInfo)
+{
+	int Version = -1;
+	if(InfoExSize >= 4)
+	{
+		Version = 0;
+	}
+	else if(InfoExSize >= 8)
```
Comment (Learath2) [inline src/game/client/gameclient.cpp]: This condition is useless as the one above gets executed `\forall InfoExSize >= 4`

File: src/game/client/gameclient.cpp [hunk 6]
```diff
@@ -953,6 +950,98 @@ void CGameClient::ProcessEvents()
...
+	{
+		Version = pInfoEx->m_Version;
+	}
+	int Flags = 0;
+	if(Version >= 0)
+	{
+		Flags = pInfoEx->m_Flags;
+	}
+	bool Race = IsRace(pFallbackServerInfo);
```
Comment (Learath2) [inline src/game/client/gameclient.cpp]: Small nit, these are all string comparisons that if we have `Version => 1` will go to waste.

File: src/game/client/gameclient.cpp [hunk 7]
```diff
@@ -953,6 +950,98 @@ void CGameClient::ProcessEvents()
 	}
 }
 
+static CGameInfo GetGameInfo(const CNetObj_GameInfoEx *pInfoEx, int InfoExSize, CServerInfo *pFallbackServerInfo)
```
Comment (Learath2) [inline src/game/client/gameclient.cpp]: What is the Fallback really for? Maybe you should have two `GetGameInfo`s, one `static CGameInfo GetGameInfo(const CNetObj_GameInfoEx *pInfoEx, int InfoExSize);` and one `static CGameInfo GetGameInfo(CServerInfo *pFallbackServerInfo);`

Also why is the size of the info object needed?

File: src/game/client/gameclient.h [hunk 8]
```diff
@@ -192,7 +224,7 @@ class CGameClient : public IGameClient
 		//const CNetObj_PlayerInfo *m_paInfoByTeam[MAX_CLIENTS];
 		const CNetObj_PlayerInfo *m_paInfoByDDTeam[MAX_CLIENTS];
 
-		const CNetObj_DDNetGameInfo *m_pGameInfoEx;
+		const CNetObj_GameInfoEx *m_pGameInfoEx;
```
Comment (Learath2) [inline src/game/client/gameclient.h]: This isn't needed anymore.

**Review by Learath2 — CHANGES_REQUESTED** [review]
Most of these are things that I want to ask rather then actual problems.

File: src/engine/client/client.cpp [hunk 9]
```diff
@@ -1531,8 +1564,29 @@ void CClient::ProcessServerPacket(CNetChunk *pPacket)
...
+			}
+			int Version = Unpacker.GetInt();
+			int Flags = Unpacker.GetInt();
+			if(Version <= 0)
+			{
+				return;
+			}
+			m_ServerCapabilities = GetServerCapabilities(Version, Flags);
+			m_GotServerCapabilities = true;
+			m_ServerSentCapabilities = true;
```
[the remark this answers is not in the document]
Comment (heinrich5991) [reply src/engine/client/client.cpp]: Renamed.

File: src/game/client/components/mapimages.h (hunk 4 above)
Comment (heinrich5991) [reply src/game/client/components/mapimages.h]: Yes, this is safe because we only assign static string literals.

**Review by Learath2 — APPROVED** [review]
LGTM

**Pull request merged**